libpromises/evalfunction.c

```diff
@@ -1124,7 +1124,25 @@ static StringSet *VariablesMatching(const EvalContext *ctx, VariableTableIterato
 
             if (pass)
             {
-                StringSetAdd(matching, expr);
+                JsonElement *data = NULL;
+                bool allocated = false;
+                if (collect_full_data)
+                {
+                    data = VarRefValueToJson(ctx, fp, v->ref, NULL, 0, true, &allocated);
+                }
+
+                if (data == NULL)
+                {
+                    JsonObjectAppendString(matching, expr, "");
```

Comment:
> When we don't collect the full variable data, we create a key-value map JsonObject with empty strings as the values. It will be destroyed soon afterwards.

This is very useful, I was hoping for a comment. :-) No big deal though, thanks for explaining.